Quick Start: Alert Appearance

nheagy opened the pull request:
Refs #9447

This improves the appearance of the quick start alerts:

![quick start alert styling](https://user-images.githubusercontent.com/517257/44074712-9edbc290-9f4e-11e8-8235-1ad947ff3144.gif)

### Tech discussion
Two changes of note:
- Method `createHighlightMessage(base: String, highlight: String, icon: UIImage) -> NSAttributedString` for creating the strings with highlights icons + words
- Enum case `quickStart(NSAttributedString?)` for passing the attributed string created by the above to the Notice

To test:
- View the Quick Start checklist for any wpcom site
- Tap an incompleted checklist item
- Ensure the notice/alert that appears at the bottom of the screen resembles the mockup

frosty commented on WordPress/Classes/ViewRelated/Blog/QuickStartTours.swift:
```diff
@@ -7,6 +7,49 @@ protocol QuickStartTour {
+
+
+extension QuickStartTour {
+    func createHighlightMessage(base normalString: String, highlight: String, icon: UIImage) -> NSAttributedString {
```
Nitpick, but the Swift API Guidelines recommend `make` instead of `create`

nheagy replied:
Ah, yes, you're totally correct!

frosty commented on WordPress/Classes/ViewRelated/System/Notices/NoticePresenter.swift:
```diff
@@ -95,14 +97,23 @@ class NoticePresenter: NSObject {
         })
     }
 
+    private func getNoticeView(for notice: Notice) -> NoticePresentationView {
```
Maybe `makeNoticeView` to be consistent with my earlier comment?

nheagy replied:
Sounds great.

frosty commented on WordPress/Classes/Stores/NoticeStore.swift:
```diff
@@ -54,7 +54,7 @@ struct Notice {
     public enum Style {
         case normal
-        case quickStart
+        case quickStart(NSAttributedString?)
```
I _think_ I'd prefer we just add an optional `formattedMessage` property to `Notice`, instead of having this on the Style 🤔 What do you think?

frosty replied:
Actually, perhaps at the same time we could rename it `attributedMessage`?

nheagy replied:
I wasn't sure. The constructor for the `Notice` gets a bit heavy. And then later we'll be creating notices with two action buttons, and I figured—maybe if I shove those somewhere else it's less ugly 😉

But it could be wrong

nheagy commented:
@frosty thanks for the review!

> What vertical positioning do we want for the icon

I was trying to follow the mockup which I thought looked like baseline alignment, but maybe I'm wrong. I'll double check!

https://scene.zeplin.io/project/5a86dcb255c808f8b2c44169/screen/5abcfa07eff05dcd5a05384c

> Should we add a shadow to the notice view

I'll leave that up to @SylvesterWilmott, as there's no shadow in the designs

SylvesterWilmott commented:
>I was trying to follow the mockup which I thought looked like baseline alignment, but maybe I'm wrong. I'll double check!

It's center aligned for these. Otherwise the icon will throw the text off centre visually

>Should we add a shadow to the notice view

There's no shadow on these due to the dark colour of the notices but @nheagy if you don't mind applying the same shadow we have on the light coloured ones and adding a screenshot here?

nheagy commented:
@SylvesterWilmott here's the notice with shadow:

![notification alert with shadow](https://user-images.githubusercontent.com/517257/44107088-1f1113a0-9fab-11e8-8414-778c2c2e48a5.png)

nheagy commented:
@frosty ok, this is ready for your re-view

![quick start alert styling v2](https://user-images.githubusercontent.com/517257/44176671-9384d380-a09f-11e8-9222-977f9b480a6b.png)

SylvesterWilmott commented:
@nheagy Yeah we should use background blur here
@frosty There will be some larger ones where the blur will be more apparent

For the icon placement - It's good that we have it centered now but I noticed it's smaller, is this just a limitation in iOS or can we align the text along a center line along with the icon at the size it was previously?

nheagy commented:
@frosty yup, that's the same way it's centered in this PR: https://github.com/wordpress-mobile/WordPress-iOS/pull/9970/files#diff-e94cfcba4d88949059e72d5cff16ad19R35

@SylvesterWilmott The icon is sized to match the mockups, or at least close: it's currently set at `16pt`. Zeplin says `13.3pt` but that was way too small, so I incremented by hand and compared to the mockups. I can use a specific size if you have it, tho in code its based on the gridicons frame, which by default is `24pt`, which was much larger than the mockup.

Revision with blurred background is incoming!

SylvesterWilmott commented:
>The icon is sized to match the mockups, or at least close

Ah okay! Maybe it was just the list icon, I guess it has a smaller perceived size. Although the icon itself in zeplin is 13.3pt the bounding box is 16pt. it's a bit fiddly to hover the cursor on the bounding box in Zeplin but it is there.

<img width="222" alt="screen shot 2018-08-17 at 09 00 24" src="https://user-images.githubusercontent.com/36532468/44254901-2044a400-a1fc-11e8-8335-5475c48c2871.png">

So 16pt is exactly right, thanks!

nheagy commented:
@frosty changes made, let me know what you think. Since the two NoticeViews are now so similar I changed it to subclassing instead of duplication. I also moved `makeHighlightMessage`.

<img width="465" alt="screen shot 2018-08-17 at 3 10 46 pm" src="https://user-images.githubusercontent.com/517257/44290911-e9888100-a22f-11e8-8485-8a94b5f547a6.png">

frosty commented on WordPress/Classes/ViewRelated/Blog/QuickStartNoticeView.swift:
```diff
@@ -1,9 +1,147 @@
+        configureForNotice()
+    }
+
+    static func makeHighlightMessage(base normalString: String, highlight: String, icon: UIImage) -> NSAttributedString {
```
I think this could probably move to a private `String` or `NSAttributedString` extension.

nheagy commented:
Thanks for the suggestions, @frosty, I implemented them almost exactly as you presented. Care for another look?

> The other complexity we'll need to handle in the future that this doesn't address is multiple buttons. Perhaps we update Notice to take an array of button titles + handlers

I agree. That design fits perfectly with the new `NoticeStyle`. 👍
